x/evm/keeper/msg_server.go

```diff
@@ -0,0 +1,104 @@
+package keeper
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	ethermint "github.com/cosmos/ethermint/types"
+	"github.com/cosmos/ethermint/x/evm/types"
+)
+
+// EthereumTx implements the Msg/EthereumTx gRPC method.
+func (k Keeper) EthereumTx(ctx sdk.Context, msg types.MsgEthereumTx) (*sdk.Result, error) {
+	// parse the chainID from a string to a base-10 integer
+	chainIDEpoch, err := ethermint.ParseChainID(ctx.ChainID())
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify signature and retrieve sender address
+	sender, err := msg.VerifySig(chainIDEpoch)
+	if err != nil {
+		return nil, err
+	}
+
+	var recipient *common.Address
+	if msg.Data.Recipient != nil {
+		addr := common.HexToAddress(msg.Data.Recipient.Address)
+		recipient = &addr
+	}
+
+	txHash := tmtypes.Tx(ctx.TxBytes()).Hash()
+	ethHash := common.BytesToHash(txHash)
+
+	st := types.StateTransition{
+		AccountNonce: msg.Data.AccountNonce,
+		Price:        msg.Data.Price.BigInt(),
+		GasLimit:     msg.Data.GasLimit,
+		Recipient:    recipient,
+		Amount:       msg.Data.Amount.BigInt(),
+		Payload:      msg.Data.Payload,
+		Csdb:         k.CommitStateDB.WithContext(ctx),
+		ChainID:      chainIDEpoch,
+		TxHash:       &ethHash,
+		Sender:       sender,
+		Simulate:     ctx.IsCheckTx(),
+	}
+
+	// since the txCount is used by the stateDB, and a simulated tx is run only on the node it's submitted to,
+	// then this will cause the txCount/stateDB of the node that ran the simulated tx to be different than the
+	// other nodes, causing a consensus error
+	if !st.Simulate {
+		// Prepare db for logs
+		blockHash := types.HashFromContext(ctx)
+		k.CommitStateDB.Prepare(ethHash, blockHash, k.TxCount)
+		k.TxCount++
+	}
+
+	config, found := k.GetChainConfig(ctx)
+	if !found {
+		return nil, types.ErrChainConfigNotFound
+	}
+
+	executionResult, err := st.TransitionDb(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	if !st.Simulate {
+		// update block bloom filter
+		k.Bloom.Or(k.Bloom, executionResult.Bloom)
+
+		// update transaction logs in KVStore
+		err = k.SetLogs(ctx, common.BytesToHash(txHash), executionResult.Logs)
+		if err != nil {
+			panic(err)
+		}
+	}
+
```

Comment:
do we need to do logging here to k.logger?

Reply:
we do log, but on the handler